src/Messages/OWSMessageSender.m

```diff
@@ -628,11 +629,17 @@ - (void)sendSyncTranscriptForMessage:(TSOutgoingMessage *)message
             // DEPRECATED - Remove after all clients have been upgraded.
             BOOL isLegacyMessage = ![message isKindOfClass:[OWSOutgoingSyncMessage class]];
 
-            NSDictionary *messageDict = [self encryptedMessageWithPlaintext:plainText
-                                                                toRecipient:recipient.uniqueId
-                                                                   deviceId:deviceNumber
-                                                              keyingStorage:[TSStorageManager sharedManager]
-                                                                     legacy:isLegacyMessage];
+            __block NSDictionary *messageDict;
+            // Mutating session state is not thread safe, so we operate on a serial queue, shared with decryption
+            // operations.
+            dispatch_sync([OWSDispatch sessionCipher], ^{
```

Comment:
Every time I see `dispatch_sync()`, I get worried about deadlock. Maybe we should start a CONCURRENCY.md which documents (for now) which threads/queues block on which threads/queues, so that we can do this kind of thing with confidence.